aws/ci_builder.rb

```diff
@@ -0,0 +1,80 @@
+require 'cdo/rake_utils'
+require_relative 'ci_build_metrics'
+module CiBuilder
+  require_relative '../lib/cdo/rake_utils'
+  extend self
+  BUILD_STARTED = 'build-started'.freeze
+
+  def build
+    Dir.chdir(deploy_dir) do
+      return 0 unless build_required?
+
+      FileUtils.touch BUILD_STARTED
+
+      git_update_count = update_repository
+
+      log_build_history(git_update_count)
+
+      display_commit_link
+
+      install_dependencies
+
+      notify_honeybadger
+
+      execute_ci_task
+    end
+  end
+
+  private
+
+  def build_required?
+    RakeUtils.git_updates_available? || File.file?(BUILD_STARTED) || !CDO.daemon
+  end
+
+  def update_repository
+    RakeUtils.git_fetch
+    count = RakeUtils.git_update_count
+    RakeUtils.git_pull if count > 0
+    [1, count].max # I think this is not needed. unless count can be a real number between 0 and 1.
+  end
+
+  def log_build_history(git_update_count)
+    log = `git log --pretty=format:"%h %s (%an)" -n #{git_update_count}`
+    File.write(deploy_dir('rebuild'), log)
+  end
+
+  def display_commit_link
+    git_revision = RakeUtils.git_revision
+    git_revision_short = GitUtils.git_revision_short
+
+    ChatClient.log "https://github.com/code-dot-org/code-dot-org/commit/#{git_revision}", message_format: 'text', color: 'purple'
+    write_build_status rack_env, git_revision_short, :start
+  end
+
+  def install_dependencies
+    RakeUtils.bundle_install
+  end
+
+  def notify_honeybadger
+    git_revision = RakeUtils.git_revision
+    Honeybadger.notify_new_release(rack_env, git_revision)
+  end
+
+  def execute_ci_task
+    status = begin
+      result = RakeUtils.rake_stream_output 'ci'
+      validate_git_revision if CDO.test_system?
+      result
+    rescue => exception
+      CDO.backtrace exception
+    ensure
+      FileUtils.rm BUILD_STARTED if File.file?(BUILD_STARTED)
+    end
+    status
+  end
+
+  def validate_git_revision
+    git_revision = RakeUtils.git_revision
+    raise 'Git revision unexpectedly changed during DTT' if git_revision != RakeUtils.git_revision
```

Comment:
I can add it back oops...

I do have a question on how this is written and I want your opinion.

Right now, I feel an inconsistency in behavior that I wanna fix at the end.



In my books this is how exceptions should be handle:

All methods should throw exceptions and not log into chat if something bad happened. Then the highest class (the orchestrator that doesn't exist yet), will catch all the issues and log them properly, and notify any system. That way the centralization of exception handling happens in one place and its consistent

The current approach:

- Log into chat
- Swallow exceptions most of the time (some action is usualkly perform, like adding to the chat.
- Behavior is repeated.


Somehow I feel like this was created because we are afraid to crash the ci_build script, but I think it should if we trust that we have the right exception handling in place.

Reply:
I do expect a certain number of inconsistencies; this script wasn't ever really _designed_, it just evolved over time based on new needs that kept arising. A design like you describe with an orchestrator certainly sounds good, but I don't have a good sense of how much work it would take to convert all of our existing error handling logic over to that model.

Reply:
once I am done it shouldn't be hard